src/ci_ex_states/excited_state_solver.cc

```diff
@@ -0,0 +1,1131 @@
+            // Combine selected determinants into total space
+            full_space.merge(PQ_space);
+            PQ_space.clear();
+        } else if ((ex_alg_ == "ROOT_ORTHOGONALIZE")) { // and i != (nrun - 1))
+            // orthogonalize
+            save_old_root(PQ_space, PQ_evecs, i, ref_root);
+            energies->set(i, PQ_evals->get(0));
+            pt2_energies[i] = sci_->get_multistate_pt2_energy_correction()[0];
+        } else if ((ex_alg_ == "MULTISTATE")) {
+            // orthogonalize
+            save_old_root(PQ_space, PQ_evecs, i, ref_root);
+        }
+        if (ex_alg_ == "ROOT_ORTHOGONALIZE" and (nroot_ > 1)) {
+            root_ = i;
+        }
+    }
+    op_ = sci_->get_op();
+    multistate_pt2_energy_correction_ = sci_->get_multistate_pt2_energy_correction();
+
+    dim = PQ_space.size();
+    final_wfn_.copy(PQ_space);
+    PQ_space.clear();
+
+    int froot = root_;
+    if (ex_alg_ == "ROOT_ORTHOGONALIZE") {
+        froot = nroot_ - 1;
+        multistate_pt2_energy_correction_ = pt2_energies;
+        PQ_evals = energies;
+    }
+
+    std::vector<int> mo_symmetry = mo_space_info_->symmetry("ACTIVE");
+    WFNOperator op_c(mo_symmetry, as_ints_);
+    if (ex_alg_ == "ROOT_COMBINE") {
+        psi::outfile->Printf("\n\n  ==> Diagonalizing Final Space <==");
+        dim = full_space.size();
+
+        for (size_t n = 0; n < nroot_; ++n) {
+            psi::outfile->Printf("\n  Determinants for root %d: %zu", n, sizes[n]);
+        }
+
+        psi::outfile->Printf("\n  Size of combined space: %zu", dim);
+
+        if (diag_method_ != Dynamic) {
+            op_c.build_strings(full_space);
+            op_c.op_lists(full_space);
+            op_c.tp_lists(full_space);
+        }
+        sparse_solver_->diagonalize_hamiltonian_map(full_space, op_c, PQ_evals, PQ_evecs, nroot_,
+                                                    state_.multiplicity(), diag_method_);
+    }
+
+    if (ex_alg_ == "MULTISTATE") {
+        local_timer multi;
+        compute_multistate(PQ_evals);
+        psi::outfile->Printf("\n  Time spent computing multistate solution: %1.5f s", multi.get());
+    }
+
+    if (add_singles_) {
+
+        psi::outfile->Printf("\n  Adding singles");
+
+        op_.add_singles(final_wfn_);
+        if (diag_method_ != Dynamic) {
+            if (sparse_solver_->sigma_method_ == "HZ") {
+                op_.clear_op_lists();
+                op_.clear_tp_lists();
+                local_timer str;
+                op_.build_strings(final_wfn_);
+                psi::outfile->Printf("\n  Time spent building strings      %1.6f s", str.get());
+                op_.op_lists(final_wfn_);
+                op_.tp_lists(final_wfn_);
+            } else {
+                op_.clear_op_s_lists();
+                op_.clear_tp_s_lists();
+                op_.build_strings(final_wfn_);
+                op_.op_s_lists(final_wfn_);
+                op_.tp_s_lists(final_wfn_);
+            }
+        }
```

Comment:
I think we can safely remove this add_singles functionality for now. It's never used, and if we eventually want to do this, I think a separate module makes more sense, i.e. a general uncontracted MRCI class.